tools/autograd/derivatives.yaml

```diff
@@ -1874,9 +1874,9 @@
 - name: log_sigmoid_forward(Tensor self) -> (Tensor output, Tensor buffer)
   self: log_sigmoid_backward(grad, self, buffer)
 
-- name: _log_softmax(Tensor self, int dim, bool half_to_float) -> Tensor
+- name: log_softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
   self: _log_softmax_backward_data(grad, result, dim, self.scalar_type())
```

Comment:
The fact that tests don't break means that autograd engine implicitly casts grad to the self type? Nice I guess...

Reply:
It's actually that `_log_softmax_backward_data` takes in "input_dtype" as an argument. So, if `self` is fp32 and `dtype` is fp64, then we have

```
_log_softmax_backward_data(grad: fp64, result: fp64, dim, self.scalar_type(): fp32)
```
So the `_log_softmax_backward_data` kernel casts from whatever grad dtype it's given to the input dtype.

Reply:
It does:
https://github.com/pytorch/pytorch/blob/f6bbecf8b503eb6562d52b66124a49b45a57f14a/torch/csrc/autograd/engine.cpp#L704-L706